src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+/*!
+ * Accurate_advance for an adaptive step control: Perform an adaptive step
+ * integration for a proposed step or a series of sub-steps within a required
+ * tolerance until the the accumulated curved path is equal to the input step
+ * length.
+ */
+CELER_FUNCTION real_type FieldDriver::accurate_advance(real_type step,
+                                                       OdeState* state,
+                                                       real_type hinitial)
+{
+    CELER_ASSERT(step > 0);
+
+    // Set an initial proposed step and evaluate the minimum threshold
+    real_type end_curve_length = step;
+
+    real_type h = ((hinitial > rel_tolerance() * step) && (hinitial < step))
+                      ? hinitial
+                      : step;
+    real_type h_threshold = shared_.epsilon_step * step;
+
+    // Output with the next good step
+    FieldOutput output;
+
+    // Performance integration
+    bool         succeeded       = false;
+    real_type    curve_length    = 0;
+    unsigned int remaining_steps = shared_.max_nsteps;
+
+    do
+    {
+        output = this->integrate_step(h, *state);
+
+        curve_length += output.step_taken;
+
+        if (h < h_threshold || curve_length >= end_curve_length)
+        {
+            succeeded = true;
+        }
+        else
+        {
+            h = std::fmax(output.value, shared_.minimum_step);
+            if (curve_length + h > end_curve_length)
+            {
+                h = end_curve_length - curve_length;
+            }
+        }
+        *state = output.state;
+    } while (!succeeded && --remaining_steps > 0);
+
+    CELER_ASSERT(succeeded);
+
+    return curve_length;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function for accurate_advance: advance for a given step and
+ * evaluate the next predicted step.
+ */
+CELER_FUNCTION auto
+FieldDriver::integrate_step(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a next proposed step
+    FieldOutput output;
+
+    if (step > shared_.minimum_step)
+    {
+        output = this->one_good_step(step, state);
+    }
+    else
+    {
+        // Do an integration step for a small step (a.k.a quick advance)
+        StepperResult result = stepper_(step, state);
+
+        // Update position and momentum
+        output.state = result.end_state;
+
+        real_type dyerr = truncation_error(
+            step, shared_.epsilon_rel_max, state, result.err_state);
+        output.step_taken = step;
+
+        // Compute a proposed new step
+        CELER_ASSERT(output.step_taken != 0);
+        output.value
+            = this->new_step_size(step, dyerr / (step * shared_.epsilon_step));
+    }
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Advance within a relative truncation error and estimate a good step size
```

Comment:
Is the 'advance' here warranted/consistent?  In contrast to `accurate_advance`, this routine does not update the state, it seems to just propose an advance.

Reply:
The state is updated through the output:
<pre>
output.state      = result.end_state;
</pre>

Reply:
Yes, but it is "just" a proposition that can be accepted or rejected by the caller (by ignoring or using the result).  That in the (current) single use, it happens (indirectly) to always be copying it is, possibly, an artefact.
```suggestion
 * Propose an advance within a relative truncation error and estimate a good step size
```

Reply:
Actually, it is not "proposing a step", but it evaluate a sub-step to advance within a given tolerance error and the result will be always used for advance_accurate.  The logic for advance_accurate is to sub-divide segments within a tolerance and move to the end of the propose step. The step_taken from "one_good_step" is guaranteed to meet the required error tolerance.